packages/airnode-node/src/evm/requests/blocking.ts

```diff
@@ -1,51 +1,112 @@
+    if (apiCall.status !== RequestStatus.Pending || !pendingWithdrawal) {
+      return { ...acc, apiCalls: [...acc.apiCalls, apiCall] };
     }
 
-    return { ...acc, apiCalls: [...acc.apiCalls, apiCall] };
+    const warningLog = logger.pend(
+      'WARN',
+      `Ignoring Request ID:${apiCall.id} as it has a pending Withdrawal ID:${pendingWithdrawal.id}`
+    );
+    const blockedCall: Request<ApiCall> = {
+      ...apiCall,
+      status: RequestStatus.Ignored,
+      // TODO: These error messages should be functions accepting id parameter and creating the error string
+      errorMessage: `${RequestErrorMessage.PendingWithdrawal}: ${pendingWithdrawal.id}`,
+    };
+
+    return {
+      ...acc,
+      logs: [...acc.logs, warningLog],
+      apiCalls: [...acc.apiCalls, blockedCall],
+    };
   }, initialState);
 
-  const updatedRequests = { ...requests, apiCalls };
-  return [logs, updatedRequests];
+  return [[...prevLogs, ...logs], { ...requests, apiCalls }];
+}
+
+/**
+ * We need to limit the requests from the same sponsor because they are performed in parallel. However, we cannot just
+ * filter based on sponsor or sponsor wallet only, because we have not yet validated which is the correct sponsor wallet
+ * for a given sponsor.
+ *
+ * Instead we need to limit the request based on (sponsor, sponsor wallet) pair.
+ */
+export function applySponsorAndSponsorWalletRequestLimit([
+  prevLogs,
+  requests,
+]: LogsData<GroupedRequests>): LogsData<GroupedRequests> {
+  const apiCallsWithSponsorId = requests.apiCalls.map((apiCall) => ({
+    ...apiCall,
+    // The "sponsor id" is a unique combination of sponsor and sponsor wallet
+    sponsorId: `${apiCall.sponsorAddress}-${apiCall.sponsorWalletAddress}`,
+  }));
+  const requestCountPerSponsorId = new Map<string, number>();
+  const allowedApiCalls: Request<ApiCall>[] = [];
+  const logs: PendingLog[] = [];
+
+  // TODO: Consider reduce or map for this as a functional approach
+  apiCallsWithSponsorId.forEach((apiCall) => {
+    if (apiCall.status !== RequestStatus.Pending) return;
+    const sponsorId = apiCall.sponsorId;
+
+    if (!requestCountPerSponsorId.has(sponsorId)) {
+      requestCountPerSponsorId.set(sponsorId, 0);
+    }
+    const sponsorRequests = requestCountPerSponsorId.get(sponsorId)!;
+
+    if (sponsorRequests < MAXIMUM_SPONSOR_WALLET_REQUESTS) {
+      requestCountPerSponsorId.set(sponsorId, sponsorRequests + 1);
+      allowedApiCalls.push(apiCall);
+      return;
+    }
+
+    logs.push(logger.pend('WARN', `Blocking Request ID:${apiCall.id} as it exceeded sponsor request limit.`));
+    const blockedCall: Request<ApiCall> = {
+      ...apiCall,
+      status: RequestStatus.Blocked,
+      errorMessage: `${RequestErrorMessage.SponsorRequestLimitExceeded}: ${apiCall.id}`,
+    };
+    allowedApiCalls.push(blockedCall);
+  });
+
+  return [
+    [...prevLogs, ...logs],
+    { ...requests, apiCalls: allowedApiCalls.map((apiCall) => omit(apiCall, 'sponsorId')) },
```

Comment:
Why not just calculate `sponsorId` inside the `forEach` loop? Doesn't seem like it needs to be added to the API calls and then removed again?

Reply:
No reason. I like this suggestion.